Python Dispatcher should respect FuncTorchBatchedDecomposition key

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #98328

Fixes https://github.com/pytorch/pytorch/issues/97425.

Python Dispatcher's resolve_key function should be equivalent to
computeDispatchTableEntryWithDebug. We added a section to
computeDispatchTableEntryWithDebug but forgot to add it to resolve_key.

This PR fixes that discrepancy.

Test Plan:
- new test

zou3519 committed:
Python Dispatcher should respect FuncTorchBatchedDecomposition key

Fixes https://github.com/pytorch/pytorch/issues/97425.

Python Dispatcher's resolve_key function should be equivalent to
computeDispatchTableEntryWithDebug. We added a section to
computeDispatchTableEntryWithDebug but forgot to add it to resolve_key.

This PR fixes that discrepancy.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "Python Dispatcher should respect FuncTorchBatchedDecomposition key"


Fixes https://github.com/pytorch/pytorch/issues/97425.

Python Dispatcher's resolve_key function should be equivalent to
computeDispatchTableEntryWithDebug. We added a section to
computeDispatchTableEntryWithDebug but forgot to add it to resolve_key.

This PR fixes that discrepancy.

Test Plan:
- new test

[ghstack-poisoned]

kshitij12345 commented on test/functorch/test_vmap.py:
```diff
@@ -1435,6 +1435,15 @@ def clone_contiguous(x):
         with self.assertRaisesRegex(RuntimeError, msg):
             vmap(lambda x: x.clone(memory_format=torch.channels_last_3d))(torch.randn(B0))
 
+    def test_decomposition_under_python_dispatcher(self):
```
I have checked locally that the test passes locally even without the fix.

I think TestVmapBase doesn't do what the below comment says. Probably because the `torch._C._functorch._is_vmap_fallback_enabled()` is True by default.

https://github.com/pytorch/pytorch/blob/a52cf3398c90d0407f77cce392173d597c981b7e/test/functorch/test_vmap.py#L1269-L1272

**NOTE:** Adding `torch._C._functorch._set_vmap_fallback_enabled(False)` right after the imports in this file leads to a bunch of failure. Will file an issue for the same.

zou3519 replied:
Thanks for catching that. I've added a manual test to check for the vmap fallback warning and verified it fails before this patch. We probably disabled the automatic assertion on the error at some point in the past and never turned it back on...

zou3519 committed:
Update on "Python Dispatcher should respect FuncTorchBatchedDecomposition key"


Fixes https://github.com/pytorch/pytorch/issues/97425.

Python Dispatcher's resolve_key function should be equivalent to
computeDispatchTableEntryWithDebug. We added a section to
computeDispatchTableEntryWithDebug but forgot to add it to resolve_key.

This PR fixes that discrepancy.

Test Plan:
- new test

[ghstack-poisoned]

kshitij12345 commented on test/functorch/test_vmap.py:
```diff
@@ -1057,6 +1057,20 @@ def f(x, y):
+        t = torch.ones(3, 3) * 5
+        with warnings.catch_warnings(record=True) as wa:
+            warnings.simplefilter('always')
+            with EnableVmapFallbackWarnings():
```
Wouldn't it be easier to just use `DisableVmapFallbackWarnings` from common_utils. The test would just fail automatically under that context without the fix.

zou3519 replied:
Thanks, that works. I forgot that that existed

kshitij12345 reviewed (approved):
LGTM! Thanks @zou3519 

(Have one minor non-blocking comment.)

zou3519 committed:
Update on "Python Dispatcher should respect FuncTorchBatchedDecomposition key"


Fixes https://github.com/pytorch/pytorch/issues/97425.

Python Dispatcher's resolve_key function should be equivalent to
computeDispatchTableEntryWithDebug. We added a section to
computeDispatchTableEntryWithDebug but forgot to add it to resolve_key.

This PR fixes that discrepancy.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge

zou3519 commented:
@pytorchbot merge -f "m1 testing is backlogged"
